File: Dynamic Programming/Largest rectangular sub-matrix whose sum is 0/main.py
from typing import List

class Solution:
    def zeroSumSubmat(self, mat):
        n = len(mat)
        m = len(mat[0])
        maxi = 0
        for i in range(n):
            tempRow = [0]*m
            for j in range(i,n):
                for c in range(m):
                    tempRow[c] += mat[j][c]
            
                mp = {}
                sumi = 0
                
                for c in range(m):
                    sumi += tempRow[c]
                    
                    if(sumi == 0):
                        maxi = max(maxi,(j-i+1)*(c+1))
                
                    if(mp.get(sumi) is not None):
                       maxi = max(maxi,(j-i+1)*(c-mp[sumi]))
                    else:
                        mp[sumi] = c
                
                if(sumi == 0):
                    maxi = max(maxi,(j-i+1)*m)

        return maxi
        



#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    ob = Solution()
    t = int(input())
    for _ in range(t):
        r = int(input())
        c = int(input())
        matrix = [[0 for j in range(c)] for i in range(r)]
        for i in range(r):
            t = [int(el) for el in input().split()]
            for j in range(c):
                matrix[i][j] = t[j]
        ans = ob.zeroSumSubmat(matrix)
        print(ans)
        print("~")

# } Driver Code Ends

File: Dynamic Programming/Largest rectangular sub-matrix whose sum is 0/main.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
    public:
        int zeroSumSubmat(vector<vector<int>>& mat) {
            int n = mat.size();
            int m = mat[0].size();
            int maxi = 0;

            for(int i = 0; i < n; i++) {
                vector<int> tempRow(m);
                for(int j = i; j < n; j++) {
                    for(int c = 0; c < m; c++) {
                        tempRow[c] += mat[j][c];
                    }
                    unordered_map<int, int> mp;
                    int sum = 0;
                    for(int c = 0; c < m; c++) {
                        sum += tempRow[c];
                        if(sum == 0) {
                            maxi = max(maxi, (j - i + 1) * (c + 1));
                        }
                        if(mp.find(sum) != mp.end()) {
                            maxi = max(maxi, (j - i + 1) * (c - mp[sum]));
                        } else mp[sum] = c;
                    }
                    if(sum == 0) maxi = max(maxi, (j - i + 1) * (m));
                }
            }

            return maxi;
        }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while(t--) {
        int n, m;
        cin >> n >> m;

        vector<vector<int>> mat(n, vector<int>(m));
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                cin >> mat[i][j];
            }
        }
        Solution ob;
        int ans = ob.zeroSumSubmat(mat);
        cout << ans << endl;

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends